incubator/hnc/test/conversion/conversion_test.go

```diff
@@ -0,0 +1,223 @@
+apiVersion: hnc.x-k8s.io/v1alpha1
+kind: HierarchyConfiguration
+metadata:
+  name: hierarchy
+  namespace: e2e-conversion-test-b
+spec:
+  parent: e2e-conversion-test-a`
+		MustApplyYAML(hierA)
+
+		// Convert
+		setupV1alpha2()
+
+		// Verify CRD conversion.
+		verifyCRDConversion()
+		// Verify the parent spec and the children status in the new version.
+		FieldShouldContain(hierCRD, nsA, hierSingleton, ".apiVersion", "v1alpha2")
+		FieldShouldContain(hierCRD, nsA, hierSingleton, ".status.children", nsB)
+		FieldShouldContain(hierCRD, nsB, hierSingleton, ".apiVersion", "v1alpha2")
+		FieldShouldContain(hierCRD, nsB, hierSingleton, ".spec.parent", nsA)
+	})
+
+	FIt("should convert HCs with allowCascadingDelete", func() {
+		// Before conversion, create namespace A with allowCascadingDelete.
+		MustRun("kubectl create ns", nsA)
+		hierA := `# temp file created by conversion_test.go
+apiVersion: hnc.x-k8s.io/v1alpha1
+kind: HierarchyConfiguration
+metadata:
+  name: hierarchy
+  namespace: e2e-conversion-test-a
+spec:
+  allowCascadingDelete: true`
+		MustApplyYAML(hierA)
+
+		// Convert
+		setupV1alpha2()
+
+		// Verify CRD conversion.
+		verifyCRDConversion()
+		// Verify allowCascadingDelete in the new version.
+		FieldShouldContain(hierCRD, nsA, hierSingleton, ".spec", "allowCascadingDelete:true")
+	})
+
+	FIt("should still have HC condition if it exists in v1alpha1", func() {
+		// Before conversion, create namespace B with a missing parent A (have to
+		// create A first and then delete it because otherwise the webhook will deny
+		// setting a non-existing namespace as parent).
+		MustRun("kubectl create ns", nsA)
+		MustRun("kubectl create ns", nsB)
+		hierB := `# temp file created by conversion_test.go
+apiVersion: hnc.x-k8s.io/v1alpha1
+kind: HierarchyConfiguration
+metadata:
+  name: hierarchy
+  namespace: e2e-conversion-test-b
+spec:
+  parent: e2e-conversion-test-a`
+		MustApplyYAML(hierB)
+		MustRun("kubectl delete ns", nsA)
+		FieldShouldContain(hierCRD, nsB, hierSingleton, ".status.conditions", "CritParentMissing")
+
+		// Convert
+		setupV1alpha2()
+
+		// Verify CRD conversion.
+		verifyCRDConversion()
+		// Verify conditions in the new version.
+		FieldShouldContain(hierCRD, nsB, hierSingleton, ".apiVersion", "v1alpha2")
+		FieldShouldContain(hierCRD, nsB, hierSingleton, ".status.conditions", "CritParentMissing")
+	})
+
+	It("should convert HNCConfig", func() {
+		// Convert
+		setupV1alpha2()
+
+		// Verify CRD conversion.
+		verifyCRDConversion()
+		// Verify default types in the new version.
+		//time.Sleep(20*time.Second)
+		// TEST PASSED:
+		RunShouldContainMultiple([]string{"v1alpha2", "Role", "RoleBinding"}, statusUpdateTime, "kubectl get", configCRD, "-oyaml")
+		// TEST FAILED: got "v1error: resource name may not be empty"
+		FieldShouldContain(configCRD, "", configSingleton, ".apiVersion", "v1alpha2")
+		FieldShouldContainMultiple(configCRD, "", configSingleton, ".spec.types", []string{"Role", "RoleBinding"})
```

Comment:
/hold

Until this issue is resolved. All other 4 tests PASSED.

Reply:
FYI the failed line 168 translates to `kubectl get hncconfigurations.hnc.x-k8s.io config  -o template --template={{.apiVersion}}`.

I also tried `kubectl get hncconfigurations config  -o template --template={{.apiVersion}}` as the cmd for this line and got the same error `v1error: resource name may not be empty`.

While I manually try either cmd, it worked and I got the expected result `hnc.x-k8s.io/v1alpha2`.

Reply:
passing an empty ns translates to `kubectl get hncconfigurations.hnc.x-k8s.io config "" -o template --template={{.apiVersion}}`

Reply:
Ah right verified by removing the "" and the test passed. Thanks!